src/handlers.cpp

```diff
@@ -2437,85 +2467,153 @@ void PushProfileTask::on_get_impus_failure(CassandraStore::Operation* op,
+  }
+  else
+  {
+    event.add_compressed_param("IMS subscription unchanged", &SASEvent::PROFILE_SERVICE_PROFILE);
+  }
+
+  event.add_static_param(RegistrationState::UNCHANGED);
+  event.add_var_param("");
 
+  if (_charging_addrs_present)
+  {
+    TRC_INFO("Updating charging addresses from PPR");
+    event.add_var_param(_charging_addrs.log_string());
+    put_reg_data->with_charging_addrs(_charging_addrs);
   }
   else
   {
-    send_ppa(DIAMETER_REQ_SUCCESS);
+    event.add_var_param("Charging addresses unchanged");
   }
+
+  CassandraStore::Transaction* tsx =
+    new CacheTransaction(this,
+                         &PushProfileTask::update_reg_data_success,
+                         &PushProfileTask::update_reg_data_failure);
+  CassandraStore::Operation*& op = (CassandraStore::Operation*&)put_reg_data;
+  _cfg->cache->do_async(op, tsx);
+
+  SAS::report_event(event);
+}
+
+void PushProfileTask::on_get_primary_impus_failure(CassandraStore::Operation* op,
+                                                   CassandraStore::ResultCode error,
+                                                   std::string& text)
+{
+  SAS::Event event(this->trail(), SASEvent::CACHE_GET_ASSOC_DEF_IMPU_FAIL, 0);
+  SAS::report_event(event);
+  TRC_DEBUG("Cache query failed with rc %d", error);
+  send_ppa(DIAMETER_REQ_FAILURE);
+}
+
+void PushProfileTask::on_get_primary_impus_success(CassandraStore::Operation* op)
+{
+  Cache::GetAssociatedPrimaryPublicIDs* get_primary_public_ids =
+    (Cache::GetAssociatedPrimaryPublicIDs*)op;
+  get_primary_public_ids->get_result(_default_impus);
+
+  if (_default_impus.empty())
+  {
+    TRC_INFO("No cached default public IDs found for private ID %s - failed to "
+             "update charging addresses", _impi.c_str());
+    SAS::Event event(this->trail(), SASEvent::CACHE_GET_ASSOC_DEF_IMPU_FAIL, 0);
+    SAS::report_event(event);
+    send_ppa(DIAMETER_REQ_FAILURE);
+    return;
+  }
+
+  // We can only look for a match if an IMS Subscription was present on the PPR.
+  if (_ims_sub_present)
+  {
+    // Attempt to find a match between any of the default ids of the sets the
+    // private id corresponds to, and the default id in the PPR. If no match is
+    // found the PPR is changing the default id (which is not permitted), so
+    // reject it. Otherwise, continue.
+    std::string new_default_id;
+    XmlUtils::get_default_id(_ims_subscription, new_default_id);
+    bool impu_match_found = false;
+    for (std::string current_default_id : _default_impus)
+    {
+      if (current_default_id == new_default_id)
+      {
+        impu_match_found = true;
+        break;
+      }
+    }
+    if (!impu_match_found)
+    {
+      TRC_INFO("The default id of the PPR doesn't match a default id already "
+               "known be belong to the IMPI %s - reject the PPR", _impi.c_str());
+      SAS::Event event(this->trail(), SASEvent::CHANGE_IMPU_PPR, 0);
```

Comment:
This SAS log would be a lot more useful if you included _impi and new_default_id to show what's wrong with the PPR.

See the SAS PR for more detail